Don't crash when we call __qualname__ on torch._C.ScriptFunction

c00w opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #147906
* __->__ #147894

We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

c00w committed:
Don't crash when we call __qualname__ on torch._C.ScriptFunction

We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

[ghstack-poisoned]

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1169,10 +1169,11 @@ def call_function(
             torch._dynamo.utils.warn_once(msg)
             unimplemented(msg)
         else:
+            qualname = self.value.__qualname__
```
Couldn't this line still crash if `__qualname__` doesn't exist?

jansel replied:
+1 I don't see how this fixes the error.

c00w replied:
Sorry was writing this in the middle of a SEV and screwed it up - Code is fixed (and a couple more callsites added).

williamwen42 reviewed (commented):
I noticed that not all of the `self.value.__qualname__`s are replaced - is that intentional?

jansel reviewed (changes requested):
Please add a test for this fix.

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
Test added for the direct function call (not very clear how to me to repro this at the top level).

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
None on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1169,14 +1169,15 @@ def call_function(
             torch._dynamo.utils.warn_once(msg)
             unimplemented(msg)
         else:
+            qualname = getattr(self.value, "__qualname__", "Unknown qualname")
```
Nit: `<unknown qualname>`

c00w replied:
done.

williamwen42 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1194,7 +1195,7 @@ def call_function(
                 if self.value.__module__ in known_python_builtin_modules:
```
Missed a `self.value.__qualname__` on L1187.

c00w replied:
Good catch (redid the search and didn't spot any more, so hopefully this is the last one).

williamwen42 reviewed (commented):
A few more nits

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
@pytorchbot merge

c00w commented:
@pytorchbot merge -f "issue is a mac runner which is not related"

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
@pytorchbot merge

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

c00w commented:
@pytorchbot merge -f "rebase due to conflict"

c00w committed:
Update on "Don't crash when we call __qualname__ on torch._C.ScriptFunction"


We've root caused this to correctly throwing attribute error on ScriptFunction
when missing attributes are caused. This PR will fix crashes that are showing
up. I'm going to stack a second PR to fix torch._c.ScriptFunction just being a
very badly behaving python object (which should also fix this

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]
